main.go

```diff
@@ -3,117 +3,169 @@ package main
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"log"
+	"net"
+	"net/http"
 	"os"
 	"os/signal"
 	"path"
 	"syscall"
 
 	"github.com/hashicorp/vault-lambda-extension/config"
 	"github.com/hashicorp/vault-lambda-extension/extension"
+	"github.com/hashicorp/vault-lambda-extension/server"
 	"github.com/hashicorp/vault-lambda-extension/vault"
+	"github.com/hashicorp/vault/api"
 )
 
 const (
 	extensionName = "vault-lambda-extension"
 )
 
-var (
-	extensionClient = extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
-)
-
 func main() {
 	logger := log.New(os.Stderr, fmt.Sprintf("[%s] ", extensionName), log.Ldate|log.Ltime|log.LUTC)
 	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
We use this `ctx` to cleanly shut down our `processEvents` function, should it also be passed into the modified `initialiseExtension` to cleanly shut down the server?

Reply:
I'm not aware of a way to use a context to directly cause an HTTP server to shutdown. However, we do handle clean shutdown in the goroutine below by calling `srv.Shutdown` when we receive an interrupt signal or a SHUTDOWN event from the runtime API.

Your comment got me looking closely though and one pre-existing place we don't handle shutdown well is we don't pass a context to the Vault login or read calls for `writePreconfiguredSecrets`. However, I'm not sure how to handle this. We can't start listening for SHUTDOWN from the runtime API, because to register for those events would signal our readiness too early. So all we could handle is receiving OS shutdown signals, but I'm not sure in what scenario we would get those? Perhaps if the runtime API wants to shutdown the extension before it has initialised? I would guess that could happen if the extension takes longer than 10s to startup.

Either way though, this PR doesn't make that bit worse, so I'll probably save those changes for another PR if that all sounds right to you.

Reply:
I'm not sure if the code changed or I just didn't read it well enough, I think my concern of the server _not_ being shut down if the interrupt or shutdown were received is at the least no longer valid, I see the `cancel()` and then `srv.Shutdown` now. I was thinking that `initialiseExtension` needed to listen for `ctx.Done` but I see now we should be good with regards to the shutdown